feat(cardano): Message signing

jaskp opened the pull request:
<!--
If you are a core dev:

Don't forget to set up the fields:
- assign yourself
- set priority to same as original issue
- add PR to sprint
- if Draft -> "in progress"
- if final PR -> "needs review"

If you're an external contributor, you can ignore this text.
-->

jaskp committed:
feat(cardano): Message signing

davidmisiak commented on python/src/trezorlib/cardano.py:
```diff
@@ -298,37 +300,46 @@ def _parse_tokens(tokens: Iterable[dict], is_mint: bool) -> List[messages.Cardan
+    if "addressType" not in output:
+        raise ValueError(error_message)
+
+    address_parameters = tools.dict_from_camelcase(output)
+    address_parameters.setdefault("path", "")
+    address_parameters.setdefault("staking_path", "")
```
I'm not quite sure what's the actual purpose of `_parse_output_address_parameters`, but if it's only converting the (arguably wrong) camel case expected on some places, I would reflect that in the name (instead of calling it "output"). On the other hand, maybe we could stick to the camel case address parameters also for message signing, like everywhere else?

I don't see a reason why the optionality of `path` and `staking_path` couldn't be handled in `parse_optional_address_parameters`.

davidmisiak commented on core/src/apps/cardano/sign_message.py:
```diff
@@ -0,0 +1,160 @@
+    if len(chunk_data) > _MAX_CHUNK_SIZE:
+        raise ProcessError("Chunk too large")
+
+    if chunk_index < chunks_count - 1 and len(chunk_data) < _MAX_CHUNK_SIZE:
+        raise ProcessError("Intermediate chunk too small")
```
Right, but in case of InlineDatums, the check that the total number of bytes is correct [is performed](https://github.com/trezor/trezor-firmware/blob/main/core/src/apps/cardano/helpers/hash_builder_collection.py#L63-L64) by `HashBuilderEmbeddedCBOR`. But I agree it's easy to miss - if we add a generic chunk manager, the correct length would be checked twice (in the chunk manager and in the hash builder) - but I think it's okay that way.

jaskp replied:
Oh, you mean checking whether the chunks add up to the specified byte size. I'll definitely add that check to the generic chunk iterator that will be in the following commits

davidmisiak reviewed (commented):
The `ChunkIterator` is an improvement I think :+1: 

Apart from the comments, there are some formatting errors reported by `make pystyle_check`.

jaskp commented on core/src/apps/cardano/helpers/chunks.py:
```diff
@@ -0,0 +1,83 @@
+    chunks_count = _get_chunks_count(total_size)
+    assert chunk_index < chunks_count
+
+    chunk_name = chunk.__class__.__name__
```
This ends up printing just `"Invalid last Msg: Size inconsistent with total bytes",`, I guess i should just use "chunk" instead of the class name?

davidmisiak replied:
I think using only `chunk` is okay.

janmazak commented on common/protob/messages-cardano.proto:
```diff
@@ -500,3 +500,49 @@ message CardanoTxBodyHash {
+ * @next CardanoMessageItemAck
+ */
+message CardanoSignMessageInit {
+    required uint32 protocol_magic = 1;                           // network's protocol magic
```
What's the purpose of this? I don't think we are doing anything with Byron addresses for message signing. And isn't network id part of `CardanoAddressParametersType`? (If not, why is it required while the address is optional?)

jaskp replied:
@janmazak Network id isn't part of CardanoAddressParametersType, but you are right in that protocol magic and network id should be as optional as address parameters.

As for byron address, our discussion about them in slack hasn't reached a conclusion yet, please chime in

janmazak commented on core/src/apps/cardano/layout.py:
```diff
@@ -276,32 +277,70 @@ async def confirm_inline_datum(first_chunk: bytes, inline_datum_size: int) -> No
+        props = _get_data_chunk_props(
+            "Empty message", payload_first_chunk, payload_size
+        )
+    elif not prefer_hex_display and is_unambiguous_ascii(payload_first_chunk):
```
I'm wary of doing any guesswork here. Why don't we let the client choose the "display mode" (ascii/hex) explicitly and just validate it's OK?

janmazak commented on core/src/apps/cardano/helpers/chunks.py:
```diff
@@ -0,0 +1,83 @@
+
+    is_last_chunk = chunk_index == chunks_count - 1
+
+    if not is_last_chunk and len(chunk.data) < MAX_CHUNK_SIZE:
```
Maybe checking this is not really needed? (Not sure e.g. because of UI.)
(We are not this strict on Ledger.)

davidmisiak replied:
Maybe we could do without the check, but since the communication is usually (always?) done by Connect, the client is not affected by the check in any way. And having a standard form is good security-wise, I think.

janmazak commented on core/src/apps/cardano/layout.py:
```diff
@@ -265,43 +266,76 @@ async def confirm_datum_hash(datum_hash: bytes) -> None:
+) -> list[PropertyType]:
     MAX_DISPLAYED_SIZE = 56
     displayed_bytes = first_chunk[:MAX_DISPLAYED_SIZE]
     bytes_optional_plural = "byte" if data_size == 1 else "bytes"
-    props: list[tuple[str, bytes | None]] = [
+    props: list[PropertyType] = [
         (
             f"{title} ({data_size} {bytes_optional_plural}):",
             displayed_bytes,
         )
     ]
     if data_size > MAX_DISPLAYED_SIZE:
         props.append(("...", None))
```
Hmm, I think it would be best if we showed the whole ASCII message even if it is long. 
For bytes in hex, showing the whole of the message does not really add security for most users --- they are likely to just be annoyed and skip it quickly. And the more paranoid users can avoid the feature (and ask us to show everything --- I doubt there will be demand for that, but we will see) or only use it when long messages are hashed.

janmazak commented on common/protob/messages-cardano.proto:
```diff
@@ -500,3 +500,49 @@ message CardanoTxBodyHash {
+    repeated uint32 signing_path = 3;                             // BIP-32-style path to derive the signing key from master node
+    required uint32 payload_size = 4;                             // size of the payload to be signed
+    required bool hash_payload = 5;                               // whether to hash the payload before signing
+    repeated uint32 key_path = 6;                                 // mutually exclusive with address_parameters
```
This field seems not really useful --- if we use key hash for the address field, it will be the key used to sign the msg, so it seems we can just use `signing_path`?

jaskp replied:
I assumed it was implied by this part of the doc that we want to separate the two paths.
> Since the address field might be unrelated to the key used for signing and it seems impossible to completely determine what the existing applications use, we don’t want to impose restrictions on the derivation paths used.

Did I understand it wrong?

janmazak replied:
The sentence you quote is somewhat ambiguous. What I meant was that in general according to CIP-8 it might be unrelated and some historical use could be there _in case of address_. The usage of just key hash in the address field is new (AFAIK --- discussed with Ryan, gitmatchtl etc.), so we do not need to support arbitrary keys different from signing keys there. I'd be OK if you kept it implemented like this (if it works now). What makes sense to me is not to display the key path twice if it is the same --- i.e. if there is just a key hash in the address field and the key is used for signing, only show that once (I'm not sure at which place, because I don't know how people use the arbitrary msg signing feature, but I'd perhaps put more emphasis on the key that is being used to sign and always explicitly say what that key path is, and hide info about the address field from the user if it does not say anything new.)

janmazak replied:
I guess I won't add "arbitrary key in addr field (different from signing key)" to Ledger because it would cut into APDU space. I'll see today/tomorrow.

janmazak reviewed (commented):
Do we always show the message hash? (Even for ascii messages that are signed raw, not as hash.) We were told that the hash is typically displayed by dapp/sw wallet, so it's useful to the users to compute and show it.

jaskp commented:
> Do we always show the message hash? (Even for ascii messages that are signed raw, not as hash.) We were told that the hash is typically displayed by dapp/sw wallet, so it's useful to the users to compute and show it.

Yes

davidmisiak commented on common/tests/fixtures/cardano/sign_message.json:
```diff
@@ -0,0 +1,119 @@
+      }
+    }
+  ]
+}
```
Please also add a long ascii message (so that we have a UI test for that).

jaskp committed:
fix(cardano): Change display_ascii flag to prefer_hex_display

This makes the behavior more consistent with Ethereum message signing.

jaskp committed:
fix(cardano): Return pubkey

This ends up being useful for software wallets.

gabrielKerekes commented on core/src/apps/cardano/sign_message.py:
```diff
@@ -0,0 +1,186 @@
+    )
+
+    return CardanoSignMessageFinished(
+        signature=signature, address=address, pub_key=node.public_key()
```
🙏 

```suggestion
        signature=signature, address=address, pub_key=remove_ed25519_prefix(node.public_key())
```

jaskp committed:
feat(cardano): Show path, address params and longer payload chunk

Show signing path, validate with keychain, show address parameters and increase max displayed bytes of first payload chunk unless signing hash.

Also add issue number to changelog.

davidmisiak committed:
chore(cardano): add missing translations

davidmisiak committed:
chore(cardano): update UI fixtures

matejcik committed:
chore(translations): normalize strings to NFC not NFKC

romanz committed:
refactor(core): step `loop.race()` task without `loop.schedule()`

[no changelog]

obrusvit committed:
feat(core): firmware suspend by power button

- power button event is handled before the event is passed to rust
layout

[no changelog]

overcat committed:
refactor(core): improve Stellar signing UX

romanz committed:
chore(core): remove THP fallback support

It will be reimplemented after event loop restart refactoring.

[no changelog]

cepetr committed:
feat(core): derive entropy from master key

[no changelog]

bieleluk committed:
feat(core): introduce props spacing

bieleluk committed:
chore(core): change string tuple to propertytype

[no chengelog]

onvej-sl committed:
feat(core/prodtest): implement tropic prodtest

onvej-sl committed:
fix(core): fixes after libtropic update

[no changelog]

onvej-sl committed:
chore(core): update ts-tvl to 2.2

[no changelog]

onvej-sl committed:
feat(core): remove `tropic.get_certificate()`

[no changelog]

TychoVrahe committed:
refactor(core): rename debug overlay to performance overlay

[no changelog]

TychoVrahe committed:
fix(core): fix eckhart performance overlay

[no changelog]

TychoVrahe committed:
refactor(core): rewrite debug overlay in rust

[no changelog]

cepetr committed:
refactor(core): simplify entropy module, relocate/rename to storage_salt

[no changelog]

cepetr committed:
refactor(core): cleanup storage syscalls, prepare early initialization

[no changelog]

cepetr committed:
refactor(core): remove circular dependencies, new 'storage' feature

[no changelog]

cepetr committed:
chore(core): remove unused define

[no changelog]

TychoVrahe committed:
refactor(core/prodtest): unify binary update function

[no changelog]

TychoVrahe committed:
feat(core/prodtest): add boardloader update command

[no changelog]

TychoVrahe committed:
feat(core/tools): make nrf update script more universal and allow updating boardloader with it

[no changelog]

TychoVrahe committed:
chore(core): remove reflash tool

[no changelog]

TychoVrahe committed:
fix(core): fix BLE pairing

[no changelog]

TychoVrahe committed:
chore(core): update T3W1 embedded binaries

[no changelog]

ibz committed:
refactor(core): make SLIP-24 UI independent of ETH

[no changelog]

ibz committed:
feat(core): use new SLIP-24 UI for Bitcoin

ibz committed:
fix(core): allow text details memos in Bitcoin

[no changelog]

ibz committed:
refactor(core): drop outdated Bitcoin SLIP-24 UI

[no changelog]

ibz committed:
feat(tests): SLIP-24 tests for Bitcoin

* replace Dash and GRS with some coins that we still support
* drop unnecessary input flow (since we will now visit all menus by
default)
* add a test for the most basic use case: Swap with refund address and
extra message
* run tests on all current models

ibz committed:
feat(core): SLIP-24 UI for Ripple

romanz committed:
test(core): allow wiping via DebugLink

It allows THP-based debug builds to be wiped without unlocking.

[no changelog]

TychoVrahe committed:
fix(core): reset primask in emergency rescue

This allows RSOD to work properly in case the crash occurs in critical sections with interrupts disabled via primask

[no changelog]

bieleluk committed:
chore(tests): vertical menu content

bieleluk committed:
chore(tests): make go to menu item an action

obrusvit committed:
feat(core): adjust wireless setup QR code

- correct link: https://trezor.io/setup/ts7
- slightly adjusted padding

[no changelog]

ibz committed:
chore(tests): update fixtures

[no changelog]

bieleluk committed:
chore(core): add default params to propslist

[no changelog]

obrusvit committed:
feat(core): adjust Eckhart screen border

- render the border 1px from the edge towards center to account for
minor irregularities in display production

[no changelog]

romanz committed:
refactor(core): move message reassembly into a separate class

Also, simplify the reassembly logic a bit.

[no changelog]

M1nd3r committed:
chore(python): move dev keys from BootloaderV2Image
[no changelog]

obrusvit committed:
feat(core): render logo on welcome screen

- also remove unused icons

[no changelog]

obrusvit committed:
feat(core): adjust RSOD

[no changelog]

obrusvit committed:
feat(core): show pairing code in big font in bld

- introduce new FontInfo for the largest available font (72) which
contains only the numerals, this is used in bootloader for the pairing
code comparison screen

[no changelog]

romanz committed:
refactor(core): introduce `ThpContext`

This PR moves `thp_main` code to a "context"-like class, similar to
`CodecContext` for Protocol v1.

It encapsulates payload reassembly, as well as handling and responding
to low-level single packet THP messages.

Also, channel cache is now using interface number for `CHANNEL_IFACE`,
in order to simplify channel interface serialization.

[no changelog]

romanz committed:
chore(core): remove logging from `alternating_bit_protocol`

[no changelog]

romanz committed:
chore(core): reduce default value in `MockHID` c-tor

It would allow to use it as `CHANNEL_IFACE` (in a following commit).

[no changelog]

romanz committed:
chore(core): fix a small typo in `thp_common.py`

Also, use `log._min_level` to suppress debug logging.

[no changelog]

ibz committed:
fix(core): Ripple unit formatting

[no changelog]

ibz committed:
fix(core): remove unnecessary changelog

[no changelog]

romanz committed:
refactor(core): simplify THP write-related code

- move fragmentation code into `PacketHeader.fragment_payload()`
- use `ThpContext.write_payload()` to send responses to the host
- precompute `INIT_HEADER_LENGTH` and `CONT_HEADER_LENGTH`
- remove most of `trezor.write.thp.writer` code

[no changelog]

ibz committed:
feat(core): SLIP-24 payment requests for Cardano

[no changelog]

jaskp committed:
chore(cardano): changes after rebase

bieleluk committed:
chore(core/eckhart): update tutorial screen

- add independent screen background component
[no changelog]

bieleluk committed:
chore(tools): warn about unsupported characters

- when 'º' is not supported, use 'o' as fallback

obrusvit committed:
chore(core): remove all fonts data

- they now live solely in a separate repo: `trezor-firmware-fonts`
- README is created with this information
[no changelog]

obrusvit committed:
chore(core): regen foreign chars

[no changelog]

romanz committed:
fix(core): don't reuse `Shutdown` exception object

Otherwise, its traceback object is growing without being cleared.

[no changelog]

jaskp committed:
feat(cardano): stop hashing in message signing

TychoVrahe committed:
chore(core/prodtest): improve nrf-update and bootloader-update description in readme

[no changelog]

jaskp committed:
feat(cardano): replace chunking with length+offset

M1nd3r committed:
chore(core): make CPace implementation more explanatory
[no changelog]

M1nd3r committed:
chore(core): change encoding of `code_code_entry` used in CPace
[no changelog]

M1nd3r committed:
chore(tests): adjust THP pairing tests to the new `code_code_entry` encoding
[no changelog]

jaskp committed:
feat(cardano): reject probable hashes

bieleluk committed:
feat(core/eckhart): haptic feedback for HTC animation

[no changelog]

M1nd3r committed:
chore(python): fix `code_entry_code` encoding in trezorlib
[no changelog]

obrusvit committed:
fix(build): no ui_debug_overlay only in prod

[no changelog]

jaskp committed:
fix: changelog formatting

romanz committed:
chore(core): remove unused `interface_manager`

[no changelog]

romanz committed:
chore(core): export `memzero()` via `trezor.utils`

[no changelog]

romanz committed:
fix(core): don't reuse `FORBIDDEN_KEY_PATH` exception object

Following https://github.com/trezor/trezor-firmware/issues/5472#issuecomment-3174382038.

[no changelog]

ibz committed:
feat(core): SLIP-24 payment requests for Solana

[no changelog]

obrusvit committed:
chore: remove forgotten diff file

- file accidentally added in b7fb51a7b0ca2b5fd96354126e12bba15cbbf27e

[no changelog]

mmilata committed:
fix(core): set model variant in THP DeviceProperties

[no changelog]

TychoVrahe committed:
chore(core): set T3W1 full name

[no changelog]

obrusvit committed:
feat(core): add Eckhart LED colors palette

- colours chosen according to Figma StyleKit

[no changelog]

bieleluk committed:
fix(core/io): Incomplete disabling of haptics

obrusvit committed:
feat(core): use LED colors for diode in Eckhart

- change affects Homescreen LED color
- newly RSOD ErrorScreen uses LED diode

[no changelog]

TychoVrahe committed:
feat(core/bootloader): start the device on short press

[no changelog]

bieleluk committed:
fix(core/tests): wrong eckhart action

TychoVrahe committed:
feat(core/bootloader): improve bootloader entry sequence on T3W1

[no changelog]

obrusvit committed:
chore(core): remove unusued code from BldHeader

[no changelog]

romanz committed:
refactor(core): move checksum validation to `Channel`

Also, move channel usage update to `interface_context`.

Keep the last valid reassembled message at `Reassembler.message`.
It will be used to implement `ThpChannel` read functionality.

[no changelog]

mmilata committed:
chore(core): expose BLE connection flags to python

[no changelog]

ibz committed:
fix(tests): add t3w1 to skiplist

[no changelog]

TychoVrahe committed:
feat(core): implement explicit secret sector locking mechanism

[no changelog]

TychoVrahe committed:
feat(core): use explicit secret locking on T3W1 and allocate space for MCU device cert in the secret sector

[no changelog]

TychoVrahe committed:
fix(core/prodtest): fix tropic version and chip id commands error handling

[no changelog]

TychoVrahe committed:
chore(core): update libtropic

[no changelog]

mmilata committed:
feat(core): extend THP credentials with app_name

[no changelog]

romanz committed:
feat(core): reimplement THP event loop restarts

[no changelog]

TychoVrahe committed:
fix(core): tropic fixes after libtropic update

[no changelog]

TychoVrahe committed:
feat(core/prodtest): implement tropic FW update

[no changelog]

TychoVrahe committed:
fix(core): initialize tropic model device in emulator

[no changelog]

kopecdav committed:
fix(core): unify power manager timings.

[no changelog]

bieleluk committed:
chore(core): add LED enable/disable functionality

- align LED colors with firmware
[no changelog]

romanz committed:
test(core): reduce the probability for `test_repeated_backup_via_device` failure

[no changelog]

ibz committed:
feat(bolt): SLIP-24 payment requests

[no changelog]

M1nd3r committed:
fix(python): fix BootloaderV2Image verify
[no changelog]

andrewkozlik committed:
feat(core): Rename otp-device-id-write to otp-device-sn-write.

TychoVrahe committed:
fix(core/bootloader): improve failed communication behavior

[no changelog]

romanz committed:
chore: bump `cachix/install-nix-action` to v31

[no changelog]

romanz committed:
test(core): wait for device availability to avoid UI tests flakiness

[no changelog]

TychoVrahe committed:
fix(core): preserve BLE peer count info during suspend

[no changelog]

romanz committed:
feat(core): add THP transport-level PING/PONG messages

[no changelog]

romanz committed:
fix(core): don't reuse `TASK_CLOSED` exception object

Following https://github.com/trezor/trezor-firmware/issues/5472#issuecomment-3174382038.

[no changelog]

M1nd3r committed:
chore(python): add ROOT signing public keys
[no changelog]

M1nd3r committed:
chore(embed): add boardloader ec/pq and secmon public keys to T3W1 model
[no changelog]

romanz committed:
refactor(python): define `MAX_RETRANSMISSION_COUNT` in `protocol_v2`

[no changelog]

romanz committed:
ci(core): disable `core_flash_size_compare:` job

Since currently it uses GitLab.

[no changelog]

romanz committed:
ci: increase `poetry install` timeout

[no changelog]

romanz committed:
test(core): send `Cancel` only on T1B1

[no changelog]

romanz committed:
test(core): wait for debuglink task before restarting THP event loop

[no changelog]

romanz committed:
test(core): allow joining debuglink task before restarting event loop

[no changelog]

obrusvit committed:
feat(core): show chrg icon in Eckhart BldHeader

[no changelog]

obrusvit committed:
feat(core): extend fuel gauge UI component

- new mode for showing only the charging icon indication

[no changelog]

bieleluk committed:
chore(core/eckhart): adapt loader for  click tests

[no changelog]

romanz committed:
fix(python): skip `EndSession` in case of invalidated THP channel

[no changelog]

romanz committed:
feat(python): ignore retransmitted THP payloads

[no changelog]

romanz committed:
refactor(python): don't return SEQ bit for irrelevant THP messages

[no changelog]

romanz committed:
refactor(python): inline `ProtocolV2Channel.prepare_channel_without_pairing()`

[no changelog]

cepetr committed:
fix(core): fix irq locking in ble driver

[no changelog]

onvej-sl committed:
style(core): make style consistent

[no changelog]

bieleluk committed:
feat(core/eckhart): add label keyboard layout

- utilize all passphrase input parameters
- add label changing device menu action
[no changelog]

bieleluk committed:
chore(core/eckhart): device menu translations

[no changelog]

romanz committed:
feat(core): join debuglink task before restarting Codec v1 event loop

[no changelog]

bieleluk committed:
chore(core/eckhart): update device menu about screen

[no changelog]

romanz committed:
feat(core): reimplement THP retransmissions

[no changelog]

romanz committed:
fix(core): don't reuse `_TIMEOUT_ERROR` exception object

Following https://github.com/trezor/trezor-firmware/pull/5523.

[no changelog]

TychoVrahe committed:
fix(core): fine tune BLE behavior after unexpected events

i.e. where the is nRF reset, we want to restart the pairing if there was paring mode before

[no changelog]

romanz committed:
chore(core): update UI fixtures for T2T1

Following https://github.com/trezor/trezor-firmware/pull/5563.

bieleluk committed:
feat(core/eckhart): regulatory screen

[no changelog]

bieleluk committed:
chore(core/eckhart): invert QR screen colors

[no changelog]

bieleluk committed:
chore(core/eckhart): enable menu separators by default

[no changelog]

bieleluk committed:
chore(core/eckhart): update apply settings label flow

- change button label and gradient
- add success screen after changing device name
[no changelog]

bieleluk committed:
chore(core/eckhart): extend header buttons

- use header builder functions where possible
- update fixtures
[no changelog]

cepetr committed:
fix(prodtest): remove extra crlf in response in non-interactive mode

bieleluk committed:
chore(core/eckhart): make paginate buttons equally sized

[no changelog]

romanz committed:
refactor(python): refactor THP-related exception types

[no changelog]

TychoVrahe committed:
chore(core): add secmon keys to trezorlib

[no changelog]

romanz committed:
chore(core): remove forgotten `log.info()` in `interface_context.py`

[no changelog]

matejcik committed:
build(core): fix dependency versions in trezor-core-tools

otherwise it's not possible to `poetry lock` when one of trezor or
trezor-core-tools is touched, because their version specs clash

TychoVrahe committed:
fix(core/prodtest): fix conditional usage of tropic testing keys

[no changelog]

TychoVrahe committed:
feat(core/bootloader): stronger haptic effect on power on

[no changelog]

cepetr committed:
fix(prodtest): fix wpc-info command output

cepetr committed:
fix(prodtest): fix secrets-certdev-write arg count check

ibz committed:
chore(cardano): rename payload

[no changelog]

romanz committed:
fix(core): correct `ticks_diff` arguments' order

[no changelog]

romanz committed:
chore(core): exclude logging on non-debug builds

[no changelog]

ibz committed:
chore(tests): update fixtures

[no changelog]

obrusvit committed:
fix: re-lock Poetry env using the version from Nix

- Nix env has 2.1.3

[no changelog]

cepetr committed:
fix(core): increase prodtest stack size

[no changelog]

romanz committed:
fix(core): increase `MAX_SUBSCREENS` a bit

Otherwise, entering the menu when there is
a connected BLE device results in a RSOD.

[no changelog]

romanz committed:
fix(core): avoid infinite THP error loop

[no changelog]

romanz committed:
chore(core): fix log formatting in `trezor/wire/thp/channel.py`

[no changelog]

romanz committed:
fix(python): correct CLI error message on uninitialized device

[no changelog]

obrusvit committed:
chore(core): update T3W1 embedded binaries

- build command: make build_bootloader TREZOR_MODEL=T3W1
BOOTLOADER_DEVEL=1 BOOTLOADER_QA=1
- then: make bootloader_hashes

[no changelog]

bieleluk committed:
chore(core/eckhart): extend device menu FwUI param list

- update result enum with new options
[no changelog]

romanz committed:
refactor(core): inline THP `PairingContext` confirmation methods

[no changelog]

romanz committed:
chore(core): remove `thp.pairing._skip_pairing_dialog()`

It is currently unused, since `should_show_pairing_dialog` is always `True`.

[no changelog]

bieleluk committed:
chore(core): add new menu length

[no changelog]

romanz committed:
chore(core): move `show_pairing_dialog()` to `trezor.wire.thp.ui` module

[no changelog]

ibz committed:
fix(cardano): stop trimming long messages

[no changelog]
